### PR TITLE
feat: support function return type derivation

### PR DESCRIPTION
**Pull request by scgkiran**

### CODE REVIEW

File: types/parser/visitor.go [hunk 1]
```diff
@@ -324,6 +338,21 @@ func (v *TypeVisitor) VisitNumericParameterName(ctx *baseparser2.NumericParamete
 }
 
 func (v *TypeVisitor) VisitNumericExpression(ctx *baseparser2.NumericExpressionContext) interface{} {
-	// TODO handle numeric expression
-	return ctx.GetText()
+	return v.Visit(ctx.Expr())
```
[the remark this answers is not in the document]
Comment (scgkiran) [reply types/parser/visitor.go:342]: Good point. This case is something like `varchar<L1+1>` is not handled yet. 
Workaround for type derivation:
```
L2 = L1 + 1
varchar<L2>
```

This is not critical and changes are non trivial, so deferring it for now. Added the todo back.

File: types/parameterized_user_defined_type.go [hunk 2]
```diff
@@ -202,3 +202,22 @@ func (m *ParameterizedUserDefinedType) ReturnType() (Type, error) {
 	}
 	return &UserDefinedType{Nullability: m.Nullability, TypeParameters: types}, nil
 }
+
+func (m *ParameterizedUserDefinedType) WithParameters(params []interface{}) (Type, error) {
```
[the remark this answers is not in the document]
Comment (scgkiran) [reply types/parameterized_user_defined_type.go:206]: added

File: types/parameterized_list_type.go [hunk 3]
```diff
@@ -64,10 +64,20 @@ func (m *ParameterizedListType) ShortString() string {
...
-	elemType, err := m.Type.ReturnType()
+func (m *ParameterizedListType) ReturnType([]FuncDefArgType, []Type) (Type, error) {
+	elemType, err := m.Type.ReturnType(nil, nil)
 	if err != nil {
 		return nil, err
 	}
 	return &ListType{Nullability: m.Nullability, Type: elemType}, nil
 }
+
+func (m *ParameterizedListType) WithParameters(params []interface{}) (Type, error) {
```
[the remark this answers is not in the document]
Comment (scgkiran) [reply types/parameterized_list_type.go:75]: added

File: types/parameterized_decimal_type.go [hunk 4]
```diff
@@ -72,11 +68,29 @@ func (m *ParameterizedDecimalType) ShortString() string {
...
+func (m *ParameterizedDecimalType) ReturnType([]FuncDefArgType, []Type) (Type, error) {
 	precision, perr := m.Precision.(*integer_parameters.ConcreteIntParam)
 	scale, serr := m.Scale.(*integer_parameters.ConcreteIntParam)
 	if !perr || !serr {
 		return nil, fmt.Errorf("precision and scale must be concrete integer parameters")
 	}
 	return &DecimalType{Nullability: m.Nullability, Precision: int32(*precision), Scale: int32(*scale)}, nil
 }
+
+func (m *ParameterizedDecimalType) WithParameters(params []interface{}) (Type, error) {
```
[the remark this answers is not in the document]
Comment (scgkiran) [reply types/parameterized_decimal_type.go:80]: added

File: types/any_type.go [hunk 5]
```diff
@@ -49,6 +49,10 @@ func (m AnyType) ShortString() string {
 	return "any"
 }
 
-func (m AnyType) ReturnType() (Type, error) {
+func (m AnyType) ReturnType([]FuncDefArgType, []Type) (Type, error) {
```
[the remark this answers is not in the document]
Comment (scgkiran) [reply types/any_type.go]: Added tests and fixed the code to make it work.

File: extensions/simple_extension.go [hunk 6]
```diff
@@ -153,11 +160,24 @@ func (a *ArgumentList) UnmarshalYAML(fn func(interface{}) error) error {
 			(*a)[i] = arg
 
 		} else if typ, ok := arg["type"]; ok {
-			(*a)[i] = TypeArg{
+			arg := TypeArg{
```
[the remark this answers is not in the document]
Comment (scgkiran) [reply extensions/simple_extension.go:164]: added

**Review by EpsilonPrime — APPROVED** [review]

**Pull request merged**